### ISSUE TITLE
Add support for multiple PDB per Pod in EvictionAPI

### ISSUE THREAD

**Issue by ichekrygin** [kubernetes/kubernetes#113891]
### What would you like to be added?

Kubernetes PDB definition uses a label selector to match pods for which PDB should be enforced. As such, it is possible to define multiple PDBs with different label selectors yet resolving to the same pod (or set of pods). The PDB is utilized primarily (exclusively) in two contexts: 

* Kubernetes scheduler preemption
* Kubernetes node draining (a.k.a "kubectl drain node")

The EvicitonAPI is applicable only during the node draining and is not invoked during scheduler preemption (I want to emphasize this point explicitly).
 
Kubernetes allows the definition of multiple PDBs that resolve to a single pod (or group of pods). Such a definition does not impact scheduler preemption; however, it has particular behavior in the EvicitonAPI.

The EvictionAPI returns an "InternalServerError(500)" when it detects a pod with a match against more than one PDB object.
 
https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/pod/storage/eviction.go#L214
```golang
if len(pdbs) > 1 {
    rtStatus = &metav1.Status{
        Status:  metav1.StatusFailure,
        Message: "This pod has more than one PodDisruptionBudget, which the eviction subresource does not support.",
        Code:    500,
    }
    return nil
}
```

Returning Error(500) implies that there shouldn't be any instances of multiple PDBs resolving to the same pod (or group of pods). If such behavior is intentional, we probably should disallow "overlapping pod" PDBs creation in the first place. However, I believe we could make an argument for a case when such "overlap" is beneficial and even desired functionality (see ## Use Case section below)

#### Proposal

Instead of returning InternalServerError(500), the EvictionAPI could process all resolved PDBs:

* Not a single PDB is violated - proceed with the pod eviction request.
* At least a single PDB is violated - reject pod eviction request and return Error(429)

The proposed multiple PDB handling aligns with the current (single) PDB handling and eliminates InternalServerError(500).

### Why is this needed?

Currently, there are cases when we want to provide a logical grouping for workloads that span a single workload artifact boundary. For example, create a PDB that spans multiple Deployments, Daemon/StatefulSets, or stand-alone pods.
 
A simple use-case example could be the following:

* A user has two deployments (red and blue) in the same namespace. 
* Each deployment has the identical "replicas: 3" definition.
* Each deployment has a matching PDB with a "minAvailable: 1" definition and a distinct label selector ("app=red" and "app=blue" respectively)
* A user wants to define another PDB policy with "minAvailable: 3" across both deployment pods, protecting that only one deployment can go down to 1 pod replica count at a time.

The above use case is not possible today due to the EvictionAPI multiple PDB-per-POD handling. A possible alternative to assuring that there is at least a single pod from either deployment is to define a single PDB spawning both deployments with "minAvailable: 4", which is less than ideal. 

The above use case demonstrates that multiple logical pod groups could require a separate PDB definition. A more concrete example is PDB support for Cassandra DB replication rings, where a single pod could belong to 2 (or more, depending on replication factor) CassandraDB replica sets. 

In summary, there could be multiple cases when a single workload pod can belong to numerous (logical or physical) application groups, which could benefit from a dedicated (per group) PDB. Enabling proposed support for multiple PDBs in EvictionAPI appears to align with an existing EvictionAPI PDB handling and the scope-limited to the EvictionAPI only.

**Comment by kow3ns** [comment]
The behavior is intentional. For prior discussion see [35318](https://github.com/kubernetes/kubernetes/issues/35318#issuecomment-255519251) and [75957](https://github.com/kubernetes/kubernetes/issues/75957#issuecomment-566745460). To summarize enabling this is non-trivial and modifies semantics of a v1 API. If we want to do this I think we need a KEP.

**Comment by ichekrygin** [comment]
Thank you for the feedback and the discussion links @kow3ns. I see those are pretty old :) - so I will comment on those here (rather than in the original respecive discussion issues). 

I start with https://github.com/kubernetes/kubernetes/issues/35318#issuecomment-255519251, which belongs to the issue that is tangentially related to this issue. The comment, however, provides some relevant feedback on PDB, precisely the second bullet point:
> - a pod may be in more than one service, but never have more than one controller; today PDB controller can't handle a pod being in more than one PDB-protected collection (and I'm not sure we want to ever allow it; the rules get dicey from a usability standpoint... it's basically an "and" of whether you're allowed by each PDB you're in, but I think this is a recipe for confusion, and race conditions debiting multiple PDBs for one eviction due to no transactions, etc.).

Unless I misunderstand the first part of this comment:
> today PDB controller can't handle a pod being in more than one PDB-protected collection (and I'm not sure we want to ever allow it;

I suspect it contains outdated information (after all, it was reported in 2016). The "PDB [Disruption] controller" is perfectly capable of handling "overlapping pods PDBs", i.e., pods that belong to different "controllers" (or controlling owners). The simplest case to verify this is by the following:

```
⋊> echo 'apiVersion: apps/v1 
           kind: Deployment
           metadata:
             labels:
               app: test
               color: blue
             name: blue
           spec:
             replicas: 3
             selector:
               matchLabels:
                 app: test
                 color: blue
             strategy: {}
             template:
               metadata:
                 labels:
                   app: test
                   color: blue
               spec:
                 containers:
                 - image: nginx
                   name: nginx
           ' | kubectl apply -f -
deployment.apps/blue created

⋊> kubectl create pdb test --min-available=1 --selector=app=test
poddisruptionbudget.policy/test created

⋊> kubectl create pdb blue --max-unavailable=1 --selector=color=blue
poddisruptionbudget.policy/blue created

⋊> kubectl get deploy
NAME   READY   UP-TO-DATE   AVAILABLE   AGE
blue   3/3     3            3           21s

⋊> kubectl get pod,pdb
NAME                        READY   STATUS    RESTARTS   AGE
pod/blue-76bff6bcc7-4vnwl   1/1     Running   0          24s
pod/blue-76bff6bcc7-m6qj4   1/1     Running   0          24s
pod/blue-76bff6bcc7-zj2v9   1/1     Running   0          24s

NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
poddisruptionbudget.policy/blue   N/A             1                 1                     9s
poddisruptionbudget.policy/test   1               N/A               2                     16s

⋊> kubectl scale deploy blue --replicas=2
deployment.apps/blue scaled

⋊> kubectl get pod,pdb
NAME                        READY   STATUS    RESTARTS   AGE
pod/blue-76bff6bcc7-4vnwl   1/1     Running   0          40s
pod/blue-76bff6bcc7-zj2v9   1/1     Running   0          40s

NAME                              MIN AVAILABLE   MAX UNAVAILABLE   ALLOWED DISRUPTIONS   AGE
poddisruptionbudget.policy/blue   N/A             1                 1                     25s
poddisruptionbudget.policy/test   1               N/A               1                     32s
```

Not only can we create multiple PDBs that overlap the same set of pods, but the Disruption controller can correctly react to the "controller" scale changes and update PDB status values accordingly.  

It is a little harder to comment on the second part of the comment:
> the rules get dicey from a usability standpoint... it's basically an "and" of whether you're allowed by each PDB you're in, but I think this is a recipe for confusion, and race conditions debiting multiple PDBs for one eviction due to no transactions, etc.).

since it does not provide specifics. 

I've yet to see any race conditions after spending some time with the modified version of the EvicitonAPI. The eviction request is an imperative command that either succeeds or fails. According to my (this) proposal, the success of the pod eviction request is based on the fact that a given pod does not have (belongs to) PDB that disallows disruption, and it is immaterial whether it is a single PDB or multiple. Moreover, in my opinion (again), it is less confusing to return either `Err:200` or `Err:429` as a result of the pod eviction request vs. `Err:500`.

**Comment by ichekrygin** [comment]
Also, I probably should've started with this:
> If we want to do this I think we need a KEP.

I am willing to create KEP as well as implement proposed change if we agree on the validity of this issue.

**Comment by ichekrygin** [comment]
The second link (https://github.com/kubernetes/kubernetes/issues/75957) is a "bullseye" as it relates to this issue. Since that issue is still open, I probably will follow up there.  👍 👍

**Comment by kow3ns** [comment]
> I suspect it contains outdated information (after all, it was reported in 2016). The "PDB [Disruption] controller" is perfectly capable of handling "overlapping pods PDBs", i.e., pods that belong to different "controllers" (or controlling owners). The simplest case to verify this is by the following:

No, this is correct see the [DisruptionController code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L819).

>  Moreover, in my opinion (again), it is less confusing to return either Err:200 or Err:429 as a result of the pod eviction request vs. Err:500.

The issue, unfortunately, isn't about the elegance of the existing API. We promoted the API to v1, and if users of the stable API have come to depend on the existing behavior, which we explicitly document, it's not okay to break them during a minor version upgrade.

If we want to change the semantics we have to be sure that we do so in a backward compatible way.

**Comment by ichekrygin** [comment]
> No, this is correct see the [DisruptionController code](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L819).

To clarify (and as supported by repo): the point is less of a single pod having multiple owners, but rather a PDB selector matches pods (set of pods) that are comprised of pods belonging to different controllers. I.E., `podA` from `deploymentA` AND `podB` from `deploymentB` .  Such a case is clearly supported, and the `expectedCount` [value is aggregated for all pods matching PDB selector across all controllers](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/disruption/disruption.go#L867)

**Comment by ichekrygin** [comment]
Marking as a duplicate of https://github.com/kubernetes/kubernetes/issues/75957
